core/src/storage/collections/btree_map/impls.rs

```diff
@@ -0,0 +1,1872 @@
+    }
+
+    fn correct_childrens_parent_links(
+        &mut self,
+        handle: &NodeHandle,
+        first: usize,
+        after_last: usize,
+    ) {
+        for i in first..after_last {
+            let h = KVHandle::new(handle.node, i as u32);
+            self.correct_parent_link(h);
+        }
+    }
+
+    fn correct_all_childrens_parent_links(&mut self, handle: &NodeHandle) {
+        let node = self.get_node(handle).expect("node must exist");
+        let len = node.len();
+        self.correct_childrens_parent_links(handle, 0, len + 1);
+    }
+
+    fn has_children(&mut self, handle: &NodeHandle) -> bool {
+        let node = self.get_node(handle).expect("node must exist");
+        node.edges() > 0
+    }
+}
+
+/// Densely stored general information required by a map.
+///
+/// # Note
+///
+/// Separation of these fields into a sub structure has been made
+/// for performance reasons so that they all reside in the same
+/// storage entity. This allows implementations to perform less reads
+/// and writes to the underlying contract storage.
+#[derive(Encode, Decode)]
+#[cfg_attr(feature = "ink-generate-abi", derive(Metadata))]
+pub(super) struct BTreeMapHeader {
+    /// The latest vacant index.
+    next_vacant: Option<u32>,
+    /// The index of the root node.
+    root: Option<u32>,
+    /// The number of elements stored in the map.
+    ///
+    /// # Note
+    ///
+    /// We cannot simply use the underlying length of the vector
+    /// since it would include vacant slots as well.
+    len: u32,
+    /// Number of nodes the tree contains. This is not the number
+    /// of elements!
+    pub(super) node_count: u32,
+}
+
+impl Flush for BTreeMapHeader {
+    #[inline]
+    fn flush(&mut self) {
+        self.next_vacant.flush();
+        self.root.flush();
+        self.len.flush();
+        self.node_count.flush();
+    }
+}
+
+/// A node in the tree.
+///
+/// Each node contains `CAPACITY` keys and values and an edges array over
+/// which children nodes can be linked. Each child has a link to its parent.
+///
+/// Each node is stored as one storage entity. This reduces storage access,
+/// since with each fetch the entire content of a node (all its elements, etc.)
+/// are fetched.
+#[derive(PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "ink-generate-abi", derive(Metadata))]
+pub struct Node<K, V> {
+    /// This node's parent node.
+    pub parent: Option<u32>,
+
+    /// This node's index into the parent node's `edges` array.
+    /// If, for example, `parent_idx = Some(2)` this refers to the
+    /// second position in the `edges` array of its parent node.
+    pub parent_idx: Option<u32>,
+
+    /// The array storing the keys for a node.
+    pub keys: [Option<K>; CAPACITY],
+
+    /// The array storing the values for a node.
+    pub vals: [Option<V>; CAPACITY],
+
+    /// The pointers to the children of this node.
+    pub edges: [Option<u32>; 2 * B],
```

Comment:
I'd replace `2 * B` with an `EDGES` constant and we should state that `CAPACITY + 1 == EDGES` must always be true.